### PR TITLE
feat: add ability to use more operators on flag values to create flag classes

### PR DESCRIPTION
**Pull request by onerandomusername**
## Summary

depends on #605

Add the ability to create flag instances from their values. For example: 
```py
intents = Intents.guilds | Intents.messages | Intents.reactions

msgflags = MessageFlags.ephemeral | MessageFlags.suppress_embeds
```

Additionally, support was implemented for adding and removing a flag inline:
```py
# please don't do actually do this, as this is not fully clear what is happening
intents = Intents.default() ^ (Intents.emojis_and_stickers | Intents.members)
```

## Checklist

- [x] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by onerandomusername** [comment]
Thoughts on if flag_value should support any other operators other than `|` and `^`? I don't think there's any benefit to supporting `&` on flag_values. 

Same goes for using an instance of BaseFlags with a flag_value instance: support `|` and `^` but none other.

**Comment by onerandomusername** [comment]
Thoughts on if flag_value should support any other operators other than `|` and `^`? I don't think there's any benefit to supporting `&` on flag_values. 

Same goes for using an instance of BaseFlags with a flag_value instance: support `|` and `^` but none other.

**Comment by shiftinv** [comment]
> Thoughts on if flag_value should support any other operators other than `|` and `^`? I don't think there's any benefit to supporting `&` on flag_values.

Yeah, no point in supporting `&`, since `a & b` would always result in either `a` or `0`. `flag_value` should only need `|`, any other operators don't really make sense.

> Same goes for using an instance of BaseFlags with a flag_value instance: support `|` and `^` but none other.

`__invert__` on `flag_value` might be useful, but that's really it.

**Comment by onerandomusername** [comment]
What I'm not sure about is how exactly to document this behavior with flag_values.

File: disnake/flags.py [hunk 1]
```diff
@@ -59,15 +62,24 @@
...
 
-class flag_value:
+class flag_value(Generic[T]):
     def __init__(self, func: Callable[[Any], int]):
         self.flag = func(None)
         self.__doc__ = func.__doc__
+        self._parent: Type[T] = MISSING
+
+    def __or__(self, other: flag_value[T]) -> T:
+        assert self._parent is other._parent  # noqa: S101
```
Comment (shiftinv) [inline disnake/flags.py:75]: this should raise a proper exception instead of asserting
same thing with the `BaseFlags` changes further down

**Comment by onerandomusername** [comment]
Added documentation for this feature in https://github.com/DisnakeDev/disnake/pull/615/commits/1c4633792bf3a968e3c99e7b5f2e4c10c6557af6

File: disnake/flags.py (hunk 1 above)
Comment (onerandomusername) [reply disnake/flags.py]: https://github.com/DisnakeDev/disnake/pull/615/commits/9f2838f94bfde5f92d6b247e36b2033cddfa8f34

**Comment by onerandomusername** [comment]
@shiftinv would you please leave a review?

File: disnake/flags.py [hunk 2]
```diff
@@ -173,13 +194,20 @@ def __ior__(self, other: Self) -> Self:
         return self
 
     def __xor__(self, other: Self) -> Self:
+        if isinstance(other, flag_value):
+            assert type(self) is other._parent  # noqa: S101
```
Comment (shiftinv) [inline disnake/flags.py:198]: TypeError

File: disnake/flags.py [hunk 3]
```diff
@@ -173,13 +194,20 @@ def __ior__(self, other: Self) -> Self:
...
         if not isinstance(other, self.__class__):
             raise TypeError(
                 f"unsupported operand type(s) for ^: '{self.__class__.__name__}' and '{other.__class__.__name__}'"
             )
         return self._from_value(self.value ^ other.value)
 
-    def __ixor__(self, other: Self) -> Self:
+    def __ixor__(self, other: Union[Self, flag_value[Self]]) -> Self:
+        if isinstance(other, flag_value):
+            assert type(self) is other._parent  # noqa: S101
```
Comment (shiftinv) [inline disnake/flags.py:208]: TypeError

File: disnake/flags.py (hunk 2 above)
Comment (onerandomusername) [reply disnake/flags.py]: https://github.com/DisnakeDev/disnake/pull/615/commits/d6124980db2cd5ce2ec7d05aac46df6f38f2d336

File: disnake/flags.py (hunk 3 above)
Comment (onerandomusername) [reply disnake/flags.py]: https://github.com/DisnakeDev/disnake/pull/615/commits/d6124980db2cd5ce2ec7d05aac46df6f38f2d336

File: disnake/flags.py [hunk 4]
```diff
@@ -157,13 +174,28 @@ def __iand__(self, other: Self) -> Self:
         self.value &= other.value
         return self
 
-    def __or__(self, other: Self) -> Self:
+    def __or__(self, other: Union[Self, flag_value[Self]]) -> Self:
+        if isinstance(other, flag_value):
+            if self is not other._parent:
```
Comment (shiftinv) [inline disnake/flags.py:179]:
```suggestion
            if self.__class__ is not other._parent:
```

File: disnake/flags.py [hunk 5]
```diff
@@ -328,6 +373,11 @@ class SystemChannelFlags(BaseFlags):
             Returns a new SystemChannelFlags instance with all enabled flags from both x and y.
             (Using ``|=`` will update in place).
 
+            .. versionadded:: 2.6
+        .. describe:: SystemChannelFlags.y | SystemChannelFlags.z, SystemChannelFlags(y=True) | SystemChannelFlags.z
```
Comment (shiftinv) [inline disnake/flags.py:377]: Maybe this should go next to the `~SystemChannelFlags.y` docs at the end of the list? I'm aware this is right below the `__or__` docs, but that might be more confusing than the other way.

File: disnake/flags.py [hunk 6]
```diff
@@ -157,13 +174,28 @@ def __iand__(self, other: Self) -> Self:
...
+    def __ror__(self, other: flag_value[Self]) -> Self:
+        if not isinstance(other, flag_value):
+            return NotImplemented
+        if self is not other._parent:
+            raise TypeError(
+                f"unsupported operand type(s) for |: flags of '{self.__class__.__name__}' and flags of '{other._parent.__name__}'"
+            )
+        return self._from_value(self.value | other.flag)
+
     def __ior__(self, other: Self) -> Self:
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/flags.py:199]: Probably, yes

File: disnake/flags.py (hunk 5 above)
Comment (onerandomusername) [reply disnake/flags.py]: https://github.com/DisnakeDev/disnake/pull/615/commits/1db2ed6c123f4fec0970aa921954ce3621184486

File: disnake/flags.py [hunk 7]
```diff
@@ -61,15 +64,29 @@
...
 
-class flag_value:
+class flag_value(Generic[T]):
     def __init__(self, func: Callable[[Any], int]):
         self.flag = func(None)
         self.__doc__ = func.__doc__
+        self._parent: Type[T] = MISSING
+
+    def __or__(self, other: flag_value[T]) -> T:
+        if isinstance(other, BaseFlags):
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/flags.py:77]: https://github.com/DisnakeDev/disnake/pull/615/commits/a83c66492f90191b180f9e6c7869334a2ad3eb48

File: disnake/flags.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/flags.py]: https://github.com/DisnakeDev/disnake/pull/615/commits/a83c66492f90191b180f9e6c7869334a2ad3eb48

File: changelog/605.feature.rst [hunk 8]
```diff
@@ -1,3 +1,6 @@
 Add support of more operators to all ``Flag`` classes. This list includes :class:`Intents` and :class:`Permissions`.
 - ``&``, ``|``, ``^``, and ``~`` bitwise operator support.
 - ``<``, ``<=``, ``>``, and ``>=`` comparsion operator support.
+- Support ``|`` operators between ``Flag`` classes and flag values.
+- Support ``~`` operator on flag values, which create a Flag instance with all except this specific flag enabled.
+- Support ``|`` operators between flag values which create a Flag instance with both flag values enabled.
```
Comment (shiftinv) [inline changelog/605.feature.rst:6]:
```suggestion
- Support ``|`` operators between flag instances and flag values.
- Support ``~`` operator on flag values, which create a flag instance with all except this specific flag enabled.
- Support ``|`` operators between flag values which create a flag instance with both flag values enabled.
```

File: disnake/permissions.py [hunk 9]
```diff
@@ -105,6 +105,11 @@ class Permissions(BaseFlags):
             Returns a new Permissions instance with all enabled permissions from both x and y.
             (Using ``|=`` will update in place).
 
+            .. versionadded:: 2.6
+        .. describe:: Permissions.y | Permissions.z, Permissions(y=True) | Permissions.z
```
Comment (shiftinv) [inline disnake/permissions.py:109]: should be moved to separate section similar to operations on other types

File: disnake/flags.py [hunk 10]
```diff
@@ -354,6 +401,17 @@ class SystemChannelFlags(BaseFlags):
...
+
+        Additionally supported are a few operations on class attributes.
+
+        .. describe:: SystemChannelFlags.y | SystemChannelFlags.z, SystemChannelFlags(y=True) | SystemChannelFlags.z
+
+            Returns a SystemChannelFlags instance with all provided flags enabled.
+
+        .. describe:: ~SystemChannelFlags.y
+
+            Returns a SystemChannelFlags instance with all flags except ``y`` inverted.
```
Comment (shiftinv) [inline disnake/flags.py:413]: This sounds somewhat ambiguous

File: disnake/permissions.py (hunk 9 above)
Comment (onerandomusername) [reply disnake/permissions.py]: https://github.com/DisnakeDev/disnake/pull/615/commits/b4b86b9f5ac6d22f21d4368c63de0ab256b8f8f5

File: disnake/flags.py [hunk 11]
```diff
@@ -157,29 +178,55 @@ def __iand__(self, other: Self) -> Self:
...
+            return self
         if not isinstance(other, self.__class__):
             raise TypeError(
                 f"unsupported operand type(s) for |=: '{self.__class__.__name__}' and '{other.__class__.__name__}'"
             )
         self.value |= other.value
         return self
 
-    def __xor__(self, other: Self) -> Self:
+    def __xor__(self, other: Union[Self, flag_value]) -> Self:
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/flags.py]: https://github.com/DisnakeDev/disnake/pull/615/commits/f331a41766bf68ac04a4f13372a31beb6554ec4f

File: disnake/flags.py (hunk 10 above)
Comment (onerandomusername) [reply disnake/flags.py]: https://github.com/DisnakeDev/disnake/pull/615/commits/2b90c39c5bd66117836d111c964dd0adadd901de

File: changelog/605.feature.rst (hunk 8 above)
Comment (onerandomusername) [reply changelog/605.feature.rst]: https://github.com/DisnakeDev/disnake/pull/615/commits/73eb75c96604216b5989c9fe1bf6ca4014cec14f

**Review by shiftinv — APPROVED** [review]

**Pull request merged**